### ISSUE TITLE
Kotlin type parsing fixes

### ISSUE THREAD

**Issue by anthonykeenan** [lukehutch/fast-classpath-scanner#175]
I have another issue after the latest update. I'm getting "Parameter number mismatch" when trying to parse a class file generated by Kotlin for an enum.

It looks like the <clinit> method has two parameters which are marked as synthetic, but when trying to generate the method signature it is failing [here](https://github.com/lukehutch/fast-classpath-scanner/blob/ab41c776299cbf2f4c40eab5b39888b07e63abe0/src/main/java/io/github/lukehutch/fastclasspathscanner/utils/TypeParser.java#L686) because there don't seem to be any param types.

I've created a test using the dodgy class file [here](https://github.com/anthonykr3/fast-classpath-scanner). 

Log file (with some additional debugging output)
[log2.txt](https://github.com/lukehutch/fast-classpath-scanner/files/1678076/log2.txt)

Let me know if you need any more info.

**Comment by lukehutch** [comment]
Thanks so much for the report and testcase, @anthonykr3. I'm traveling right now, but I'll try to get to this in the next few days.

**Comment by lukehutch** [comment]
@anthonykr3 -- You're right, the method type signature (stored in the extended attributes for the method) is `()V`, but in the method definition itself, the signature is defined as `(Ljava/lang/String;I)V`. I'm not sure of the right thing to do here -- should I drop the synthetic parameters when reporting the method info? Is it only synthetic parameters that may be added automatically in this way? (There are also "mandated" parameter types, which are implicit...) I'm wondering how to combine these two signatures, since sometimes I need the signature in the extended attributes, e.g. to get generic type information...

**Comment by lukehutch** [comment]
PS for reference, the problem is happening in `ClassfileBinaryParser.java`, lines 951-955:

```java
                    } else if (constantPoolStringEquals(attributeNameCpIdx, "Signature")) {
                        // Add type params to method type signature
                        final String methodTypeSignature = getConstantPoolString(readUnsignedShort());
                        methodTypeDescriptor = methodTypeSignature;
                    }
```

The `methodTypeDescriptor = methodTypeSignature;` line overwrites `(Ljava/lang/String;I)V` with `()V`. I made the assumption that `methodTypeSignature` would always contain richer type information than `methodTypeDescriptor` (e.g. generic type parameters), but it turns out that's not always the case.

**Comment by Alexander--** [comment]
I have had a look at JVM spec and Kotlin source code, but failed to understand what's going on.

The only possible hint might be in this commit: https://github.com/JetBrains/kotlin/commit/ecec87cbc75fb5727afe9a280bb876ed81cdf67f, but it looks like this specific effect (weird constructor signature) was a side effect of that change, rather than their goal.

Of course, enums aren't meant to be instantiated directly, so the constructor signature is largely irrelevant — Java even goes as far as explicitly prohibiting instantiating it's enums via reflection (AFAIK, reflection code internally checks if the target constructor belongs to enum before permitting to call it).

As for how to combine the signatures — that depends on goals of FCS users. If they want to call a method via reflection, they will obviously need JVM-friendly method descriptor. If they are using FCS to generate source code within annotation processor, they will want compiler-friendly type signature.

**Comment by lukehutch** [comment]
It's always better to provide extra information rather than throw it away.. I guess users can just discard synthetic and mandated types of that wasn't the human-visible API.

I'll try to write some code that combines the signatures in a smart way, by aligning the parameter lists according to the modifier bits, but ugh, that's ugly to have to do.

**Comment by anthonykr3** [comment]
Just a FYI - I hacked a version together today that ignored synthetic parameters (just to see whether that would be the end of my problems), and I got another error with a method that had parameters with a modifier of 0x8000, which I believe is mandated.

Then I tried hacking a version together which ignored synthetic and mandated parameters, and I found another case where there was a mandated parameter that did have param type information, so just ignoring them isn't an option.

I will try and do a PR with the updated test cases tomorrow, but thought I'd mention it in case you happened to have a look in the meantime.

**Comment by lukehutch** [comment]
@anthonykr3 Thanks for working on this. I didn't want to duplicate the work you already did, but it seems that the full solution is complex, since you really have to merge the two type signatures together for some usecases, and keep them separate for others. I took @Alexander-- 's suggestion of keeping the type signatures separate, and delivering both through the API, but I wanted to merge the type parameters from the programmer-visible type signature with the synthetic and mandated flags of the JDK-internal type signature for the `toString()` method too. I put together a patch that does this, and got your unit test to pass. Please take a look at the above patch, and let me know if it solves your problem. Thanks!

**Comment by lukehutch** [comment]
@Alexander-- thanks for the sleuthing to track down the Kotlin code involved. It's possible that Kotlin is the only language that has different parameter lists in the two type signatures -- this stuff isn't explained well anywhere I have found. It seems that synthetic and mandated parameters were added in JDK8+, and only if you use those are the parameter accessor flags even added to the method attributes block in the classfile. Anyway, I had to make a bunch of assumptions about how this worked, and the code in `TypeParser.merge()` (which merges together the two type signatures) throws `IllegalArgumentException` if any assumptions are violated, so hopefully it will be obvious if this needs to be refined further in the future.

**Comment by lukehutch** [comment]
@anthonykr3 I'm relatively confident that this patch gets us at least closer to the goal, if not solving the problem, so I released 2.12.1 so it's easier for you to test the changes. Let me know how it goes.

**Comment by anthonykr3** [comment]
Sorry, I've been testing it this morning - I've just created a pull request with another failing test, I was just working through the test case to make sure I didn't miss anything, but I'll let you have a look now

**Comment by anthonykr3** [comment]
With this one I was getting a Parameter arity mismatch between access flags and internal param types

**Comment by anthonykr3** [comment]
Which I believe is to do with the fact that it's not sending through any parameter access flags

**Comment by lukehutch** [comment]
@anthonykr3 I tried your new testcase. But you are just manually passing through `[]` as `parameterAccessFlagsInternal` -- I need the actual Kotlin class that calls the API this way, so that I can see how the classfile format is causing this API call. Can you please check into the `resources` directory a classfile that triggers the issue, as you did previously? Thank you.

**Comment by anthonykr3** [comment]
I tried that originally but it didn't work this time, the scanner just didn't find any classes in it. I'm assuming because it needed it's dependencies? So I worked out what parameters were being sent through hoping that would be enough. I've pushed another commit that has a zip file with the class file, if that helps? If not I can get the whole jar.

**Comment by lukehutch** [comment]
Try calling `.verbose()` before `.scan()` to see why the classfile isn't being found. That can happen, for example, when the class is not in the right path relative to the scanning root (usually the root directory) of the jar. I'll take another look once you are able to check in a jarfile that triggers the exception. Thanks!

**Comment by anthonykr3** [comment]
*sigh* I'd got a typo in the jar file name. I've pushed another commit where the test does fail, and it uses the jar file.

**Comment by lukehutch** [comment]
@anthonykr3 OK, what's going on is that synthetic parameters are listed in only one type descriptor, whereas mandated parameters are listed in both. I *really* hope this is not only true in Kotlin, i.e. if this is different in other JVM languages, then this is going to be very tricky to handle in a reliable way.

The modifier bits specify:

* 0x1000 (ACC_SYNTHETIC): Indicates that the formal parameter was not explicitly or implicitly declared in source code, according to the specification of the language in which the source code was written (JLS §13.1). (The formal parameter is an implementation artifact of the compiler which produced this class file.)
* 0x8000 (ACC_MANDATED): Indicates that the formal parameter was implicitly declared in source code, according to the specification of the language in which the source code was written (JLS §13.1). (The formal parameter is mandated by a language specification, so all compilers for the language must emit it.)

**Comment by lukehutch** [comment]
@anthonykr3 This should be fixed in 2.12.2. Please let me know if you run into more problems.

**Comment by anthonykr3** [comment]
So I definitely got further this time - but I've had another error. I've raised another PR with another test case. This time it looks like it's to do with the type of the parameters rather than the annotations.

Am happy to raise a separate issue if the thread is getting confusing!

**Comment by lukehutch** [comment]
@anthonykr3 This one was a type erasure issue, and thanks again for this -- each report you send makes FCS better :-)  Fixed (hopefully!) in 2.12.3.

**Comment by lukehutch** [comment]
@anthonykr3 The latest fixes are in 2.12.4. Let me know how this goes.

**Comment by anthonykr3** [comment]
Thanks @lukehutch I'll have another test.

**Comment by lukehutch** [comment]
The latest fixes are in..

**Comment by anthonykr3** [comment]
The latest version seems to be working fine for me now - I'll do some more testing on Monday but hopefully this should be it. Thanks so much for all your help though!

**Comment by lukehutch** [comment]
@anthonykr3 Great, thanks for testing. I checked in a few more changes that make type checking a bit more stringent (it requires type parameters like `<X extends xyz.Cls>` to match concrete classes after type erasure, e.g. `List<X>` should match `List<Cls>`). This doesn't break any of the tests, but please let me know if any issues result. Released in 2.12.6.

**Comment by lukehutch** [comment]
@Alexander-- @GedMarc any input you can give on the previous comment above would be much appreciated.

**Comment by lukehutch** [comment]
FYI here are the cases from @anthonykr3 's test cases so far that have mismatches between the fields described above:

```
class name: net.corda.client.jackson.StringToMethodCallParser$ParsedMethodCall
  method name: <init>
  method type descriptor: (Lnet/corda/client/jackson/StringToMethodCallParser;Ljava/lang/Object;Ljava/lang/reflect/Method;[Ljava/lang/Object;)V
    type descriptor method params: net.corda.client.jackson.StringToMethodCallParser, java.lang.Object, java.lang.reflect.Method, java.lang.Object[]
      type descriptor method param count: 4
  method signature: (TT;Ljava/lang/reflect/Method;[Ljava/lang/Object;)V
    type signature method params: T, java.lang.reflect.Method, java.lang.Object[]
      type signature method param count: 3
  parameter names: this$0, target, method, args
    parameter name count: 4
  parameter annotations: [@org.jetbrains.annotations.Nullable], [@org.jetbrains.annotations.NotNull], [@org.jetbrains.annotations.NotNull]
    parameter annotation count: 3
  parameter access flags: 0x8000, 0x0000, 0x0000, 0x0000
    parameter access flag count: 4
class name: net.corda.core.node.services.vault.GenericQueryCriteria$ChainableQueryCriteria$AndVisitor$DefaultImpls
  method name: visit
  method type descriptor: (Lnet/corda/core/node/services/vault/GenericQueryCriteria$ChainableQueryCriteria$AndVisitor;Lnet/corda/core/node/services/vault/BaseQueryCriteriaParser;)Ljava/util/Collection;
    type descriptor method params: net.corda.core.node.services.vault.GenericQueryCriteria$ChainableQueryCriteria$AndVisitor, net.corda.core.node.services.vault.BaseQueryCriteriaParser
      type descriptor method param count: 2
  method signature: <Q::Lnet/corda/core/node/services/vault/GenericQueryCriteria<TQ;-TP;>;P::Lnet/corda/core/node/services/vault/BaseQueryCriteriaParser<TQ;-TP;-TS;>;S:Lnet/corda/core/node/services/vault/BaseSort;>(Lnet/corda/core/node/services/vault/GenericQueryCriteria$ChainableQueryCriteria$AndVisitor<TQ;-TP;-TS;>;TP;)Ljava/util/Collection<Ljavax/persistence/criteria/Predicate;>;
    type signature method params: net.corda.core.node.services.vault.GenericQueryCriteria$ChainableQueryCriteria$AndVisitor<Q, ? super P, ? super S>, P
      type signature method param count: 2
  parameter names: parser
    parameter name count: 1
  parameter annotations: [@org.jetbrains.annotations.NotNull], []
    parameter annotation count: 2
  parameter access flags: 0x0000
    parameter access flag count: 1
class name: net.corda.core.contracts.ComponentGroupEnum
  method name: <init>
  method type descriptor: (Ljava/lang/String;I)V
    type descriptor method params: java.lang.String, int
      type descriptor method param count: 2
  method signature: ()V
    type signature method params: 
      type signature method param count: 0
  parameter names: $enum$name, $enum$ordinal
    parameter name count: 2
  parameter annotations: null
  parameter access flags: 0x1000, 0x1000
    parameter access flag count: 2
class name: net.corda.core.node.services.vault.AttachmentSort$AttachmentSortAttribute
  method name: <init>
  method type descriptor: (Ljava/lang/String;ILjava/lang/String;)V
    type descriptor method params: java.lang.String, int, java.lang.String
      type descriptor method param count: 3
  method signature: (Ljava/lang/String;)V
    type signature method params: java.lang.String
      type signature method param count: 1
  parameter names: $enum$name, $enum$ordinal, columnName
    parameter name count: 3
  parameter annotations: [@org.jetbrains.annotations.NotNull]
    parameter annotation count: 1
  parameter access flags: 0x1000, 0x1000, 0x0000
    parameter access flag count: 3
```

**Comment by anthonykr3** [comment]
FYI - these classes have been generated using version 1.1.60 of Kotlin. We are just in the process of upgrading to the most recent version. When I get chance I’ll re run the scanner after compiling with the most recent version of Kotlin and see whether there is any difference.

**Comment by Alexander--** [comment]
@lukehutch 

Are there any actual problems, caused by such files? There is already an API for reporting scanning errors, so as long as FCS can keep scanning, isn't everything fine?

**Comment by lukehutch** [comment]
@Alexander-- I just checked in the debug print statement code -- search `ClassfileBinaryParser.java` for `"// DEBUG PRINT STATEMENTS:`.  You can uncomment this and recompile to see the output above for classes with mismatched method parameter arity between the various fields.

Yes, there are problems caused by such files. You can't obtain full type signature information for methods (i.e. everything that is not stripped away by type erasure) at the same time as obtaining parameter annotations, parameter names, and/or parameter modifier bits (including modifiers for final, synthetic and mandated).

Out of those, parameter names and modifier bits are new in JDK8, and you have to manually compile to switch them on, unless they are generated automatically by the compiler, as they are in Kotlin (albeit buggily). But not being able to simultaneously support parameter annotations and generic type parameters for methods is a big problem for all legacy Java code in which both these functionalities are needed.

Fortunately it will probably be mostly legacy Kotlin classfiles that will be affected, and not Java classfiles, but I have no way to tell for sure without testing on some Java classfiles that exercise the type system classfile encoding as well as the Kotlin classfiles provided by @anthonykr3.

BTW it's not just a single bug we're talking about, it appears there are multiple bugs in the generation of multiple class types in Kotlin.

**Comment by lukehutch** [comment]
@anthonykr3 @Alexander-- I created this bug in the Kotlin bug tracker to track this issue:

https://youtrack.jetbrains.com/issue/KT-22706

**Comment by Alexander--** [comment]
@lukehutch 

> I can probably write a diff-like matcher that will try to do the right thing if the above conditions don't hold,…

This sounds like a bad idea. If I were you, I'd just repeat the same steps, used by javac to establish it's view of element signatures, log all inconsistencies, and call it a day.

People can do weird things with metadata as long as this does not break *their* workflow — for example, imagine creating jar with API stubs, where classes have minimal/broken type descriptors, and all type information is stored in `Signature`. As long as such class file can be used by javac to compile stuff against it, it IS a legitimate class file. It might not be useful at run time or follow spec to the letter, but it works for it's own purpose.

This reminds me about tough relationships between upx and binutils — developers of former technically follow ELF spec to the letter — when that suits them — and strip away all "debugging" metadata, not used by real world OSes (e.g. FreeBSD and Linux executable loaders); meanwhile developers of binutils generally expect, that executables *must* follow their unwritten assumptions: have full section metadata and import tables. Results: you can't debug upx-compressed executables with GDB.

**Comment by lukehutch** [comment]
@Alexander-- that's a bit of an oversimplification, since the JRE works totally fine with type erasure. FCS is a lightweight introspection system, so whatever information it returns *should* perfectly match whatever the runtime thinks is the API, but FCS can provide additional information, such as type parameters.

Actually the lack of guarantees of the JRE requiring (or enforcing) that each of the method parameter attribute types should have the same arity is a side effect of bolting on generics after Java was already in widespread use. I wish there were a few more constraints specified.

As it is, I can't even guarantee that I can return both generic type parameters and parameter annotations, even when both are present in the classfile, since one may have the same arity as the type signature, the other may have the same arity as the type descriptor, and the type descriptor and type signature may have different arity.

The other problem is that, taken as a whole across all JVM languages, I don't even know whether it is always both synthetic and mandated parameters that are missing from the type signature relative to the type descriptor, or if some languages have mandated parameters in both, but only synthetic parameters in the type descriptor. This is not defined in the spec, from what I could see, so I'm guessing that there are differences between classpath type formats found in the wild.

**Comment by lukehutch** [comment]
@Alexander-- one other point -- yes, there is a system for logging issues during scanning. But type parsing happens lazily, and as such, happens outside the main try-catch block that exceptions are logged in -- in fact, a type signature may be parsed after the log has been dumped out and finalized. For most users, the chance of something going wrong is in the `MethodTypeSignature#toString()` method, and it's not good form to throw an exception in `toString()` (but that's where these exceptions were being thrown, because consistency checks are failing).

The latest code I checked in just relies on the type signature over the type descriptor, and throws away anything that's not consistent with the type signature -- so exceptions are no longer thrown -- but this is not an ideal solution.

**Comment by GedMarc** [comment]
I've spoken to a few guys, unfortunately this is a little bit over my head (for now)... Not really a Kotlin guy (Not an anything runtime class manipulation guy, what I code I want to debug...)

So this is what I got as responses..
1) From what I hear, Kotlin also allows dynamic class manipulation at any point in the application as well - Using RefactoringHelper classes. So at any point in the application the class may or may not be structured the same as it was on any previous call. For me this brings in a lot of questions of what the good of a class scanner in such an environment would actually prove? Maybe to compare the changes to class structure, but if anything is allowed with regards to class structure, can you really build constraints around what you can read? I mean you get HotSwap which at least is still built externally to the jvm, and then you get this which is essentially runtime class hotswap. This is called the KClass.

2) Dynamic Types - Kotlin supports dynamic types (i'm guessing to mimic .net?) at which point the class is refactored to alter the variable type to the new returned type. Great for JSON returns i guess, but I like strictly typed classes. This is done utilizing point (1), so any scan done over classes and fields utilizing this feature will always be erroneous at best?
https://kotlinlang.org/docs/reference/dynamic-type.html

3) A lot of frameworks are hitting the same brick wall. I can see a lot of problems logged with Kotlin cp scanning for JUnit5, Tomcat, and a hell of a lot of others. But for me looking at (1) and (2), It definitely appears as the Kotlin language is actually built and operates in a way that a cp scan might actually be completely pointless considering the results will never actually be predictable?

Look I don't know, I personally would never use Kotlin for anything but Android development (and that is only because google recommends? it - even then I have a lot of stoppers), but its a personal gripe I suppose, and even then you need to ask the question if cp scanning JME is even valid or if there is a better way to do it (like hitting the server for the latest class versions and using the good 'ol db class versioning pattern)

Sorry guys not my area of expertise, Not sure if I'll even jump on the Kotlin bandwagon ever...

**Comment by lukehutch** [comment]
@GedMarc that's all good info, and frankly a little shocking, especially the part about type rewriting for runtime type conformity...

I guess anybody who actually uses FCS will understand that the goal of classpath scanning is to scan classfiles on disk, not whatever the latest dynamically-loaded version is in the classloader. If the type definition is morphing in RAM, the classfile on disk is not going to change, and users should expect that. The most common users will simply be looking for classes with a specific annotation, or similar.

The Kotlin bug I filed has been assigned to someone in the Kotlin bug tracker, so maybe we'll see some headway made on that end of things. At some point I'll write a more sensible fallback than the simple fallback that I have checked into the code now.

**Comment by cvmocanu** [comment]
@GedMarc

I am using Kotlin for a while and I find your comments hard to believe.

> 1) From what I hear, Kotlin also allows dynamic class manipulation at any point in the application as well - Using RefactoringHelper classes

I doubt this is possible. Do you have some pointers to some kind of documentation or example?

> 2) Dynamic Types

The [official documentation](https://kotlinlang.org/docs/reference/dynamic-type.html) says "The dynamic type is not supported in code targeting the JVM".
I hope it stays that way!

**Comment by lukehutch** [comment]
It looks like the Kotlin bug tracker bug I filed hasn't been touched yet...

**Comment by GedMarc** [comment]
@cvmocanu 
I did say not very experienced and that I did have to ask, however

https://youtrack.jetbrains.com/issue/KT-10440 shows the efforts required to get the final class actually in operation, 
Also, On this stack overflow (https://stackoverflow.com/questions/34340450/how-to-get-a-kotlin-kclass-from-a-package-class-name-string), the efforts to actually get the from a JVM or JVM environment seem pretty painful.

I really do hope you're right though!, It would be fantastic to be able to reflect kotlin from source code into it's final structure, or if there is a way, would definitely like to hear about it :)

**Comment by lukehutch** [comment]
@anthonykeenan @Alexander-- @cvmocanu @GedMarc:
I decided to fix this by right-aligning (front-padding) any of the metadata lists for method parameters that were missing values (out of the param types in the type signature, param names, param modifiers, and param annotations. (Param types in the type descriptor should have the correct arity.) This was motivated by the observation that synthetic and mandated types are _typically_ listed first in the parameter list. I haven't seen a violation of this yet, and it's the only remotely reliable way to resolve the issues illustrated in the comment above that show all of Kotlin's inconsistencies.

I also introduced a new `MethodParameterInfo` type that holds the correctly-aligned metadata for each method parameter.

Marking this as fixed; this will be in the next release, 2.19.0, which should be out in the next 24 hours. Thanks for all your comments on this bug!
